docs/development/design/watch-api.md

```diff
@@ -188,4 +188,10 @@ type Subscription interface {
 
 Since this interface does not hold events and even coalesces multiple Notify calls, it results in always being non-blocking to the callers of Notify and always having a bounded amount of memory that it uses.
 
-We have also implmented a distributed version of the SignalBus interface using the [LISTEN/NOTIFY](https://www.postgresql.org/docs/current/sql-notify.html) postgresql SQL statements.  This allows the **apiserver** to still be able to notify watch requests in other processes and thus safely scale the number of apiserver replicas past 1.
\ No newline at end of file
+We have also implmented a distributed version of the SignalBus interface using the [LISTEN/NOTIFY](https://www.postgresql.org/docs/current/sql-notify.html) postgresql SQL statements.  This allows the **apiserver** to still be able to notify watch requests in other processes and thus safely scale the number of apiserver replicas past 1.
+
+#### Scaling up large Organizations
+
+As you you start watching for change events in large organization's with many devices, it becomes harder to scale the apiserver and the backing SQL database.  This is because when a device record is changed, every connected device will have an active connection requesting those event changes and so the API server will wake up each idle connection in the long poll, and have it poll the database to get those changed device records.  Let us say you have an organization with 10,000 connected devices.  When one device is changed, that will cause 10,000 SQL queries to be executed to find out what has changed.
+
+To reduce the load on the database, we have implemented a per organization device tail cache. When a client now requests an event stream of device changes, it will execute SQL queries against the database until it reaches the tail of the change stream (this occurs when the DB query is an empty result set).  At this point now it will look for the device change record a ring buffer that is shared with all API calls looking for device changes for the same organization.  When one of the clients reaches the tail of the ring buffer, it fill the ring  with more device changes by polling the database once again. Note that only 1 is allowed to fill at a time.  If the other clients can't keep up with the fill speed of the ring buffer it, will fall back to polling the database of change data and only go back to using the ring buffer once it reaches the tail of the change stream again.  This avoids stalling event stream consumers with slow event stream consumers and caps the amount of memory the tail cache can use to be proportional to the ring buffer size.
```

Comment:
No. If you see an easy way to do it I'm all for it.

Reply:
Some other ideas to consider:

- interesting reading when thinking about in-memory options in our apiserver -- https://discuss.dgraph.io/t/the-state-of-caching-in-go-dgraph-blog/4157/9
- we could use Redis since we already have it

Reply:
Yep tend to agree.. I think we need more options here.  Think we can access the the different impl via a simple interface.  That way we can test out the options and see what's best.